design/design-doc-flexible-claims-and-composition.md

```diff
@@ -0,0 +1,945 @@
+* Allow a resource claim to dynamically provision and bind to more than one
+  managed resource, for example a `GKECluster` and a `NodePool`.
+* Limit unnecessary API churn, especially to `v1beta1` managed resource APIs
+  which must maintain backward compatibility.
+* Maintain the high-fidelity, composable nature of managed resources.
+* Maintain support for both static and dynamic provisioning.
+* Follow [declarative configuration best practices] and Kubernetes [API
+  conventions].
+* Empower the separation of concerns between infrastructure and application
+  owners.
+
+_How_ users will define resource claim kinds is out of scope for this design.
+Defining a new Kubernetes resource kind typically involves authoring a [Custom
+Resource Definition] (CRD). A method to simplify and constrain this process is
+desirable, and may be explored in future.
+
+While the design put forward by this document must allow the composition of
+managed resources, it's important to note the aim is not to enable _arbitrary
+composition_. Rather, the goal is to remove the limitations put forward in the
+[background] of this document by allowing multiple _primitive_ managed resources
+to be composed into a _composite_ managed resource. That is to say, a composite
+managed resource would conceptually be _one resource_, like "a Kubernetes
+cluster" or "an SQL instance". The design must allow an arbitrary selection of
+primitive managed resources to be composed into a composite resource, but is not
+intended to address use cases such as composing "all east coast production
+infrastructure" or "all the infrastructure needs of an application". Composite
+managed resources will be useful binding targets for resource claims, which
+represent a specific infrastructure need of an application, such as "a MySQL
+instance".
+
+## Proposal
+
+This document proposes Crossplane introduce three new kinds:
+
+* A `ResourceClass` configures how an arbitrary kind of resource claim can be
+  used to dynamically provision a single primitive managed resource.
+* A `CompositeResourceClass` configures how an arbitrary kind of resource claim
+  can be used to dynamically provision a composite managed resource, consisting
+  of multiple primitive managed resources.
+* A `CompositeResource` represents multiple primitive managed resources that
+  have been conceptually composed into a single, composite resource.
+
+Note that this document frequently uses the terms _primitive_ and _composite_ to
+describe managed resources. These terms are used in the sense of [primitive] and
+[composite] data types. Primitive managed resources are those defined, in code,
+by a Crossplane infrastructure stack. A composite managed resource can be
+thought is conceptually a single managed resource, but one that is composed from
+multiple primitive managed resources.
+
+### Example: Google Cloud SQL Instance
+
+In this example a `ResourceClass` configures how to dynamically provision a
```

Comment:
It seems like we are able to do this via `CompositeResourceClass` with only difference that `ResourceClass` has `base:` but `CompositeResourceClass` has an array like `bases: `. I'm not sure whether we need `ResourceClass` as separate kind.

> > I'm assuming that the existing resource classes like CloudSQLInstanceClass will be deprecated and eventually removed if this design works out. Do you think this will still be important if resource classes never exist as distinct CRs?

If we'd like to keep people able to provision a single primitive in an easier way compared to `CompositeResourceClass`, then I think we can keep the `CloudSQLInstanceClass` types instead of having people write `ResourceClass`. But if we'd like to standardize on `CompositeResourceClass` then any other `*Class` seems unnecessary.

Reply:
`CompositeResourceClass` and `ResourceClass` are no longer part of this design. RIP resource classes. ;)